// web-component-analyzer/lib/esm/analyze/flavors/js-doc/js-doc-flavor.js
import { parseDeclarationCSSProps } from "./parse-declaration-css-props";
import { parseDeclarationEvents } from "./parse-declaration-events";
import { parseDeclarationMembers } from "./parse-declaration-members";
import { parseDeclarationSlots } from "./parse-declaration-slots";
import { visitComponentDefinitions } from "./visit-component-definitions";
/**
 * JsDoc Flavor.
 * This flavor only looks at jsdoc and continues the visiting flavors in children nodes after emitting members.
 */
export class JsDocFlavor {
    visitComponentDefinitions(node, context) {
        visitComponentDefinitions(node, context);
    }
    parseDeclarationMembers(node, context) {
        return parseDeclarationMembers(node, context);
    }
    parseDeclarationEvents(node, context) {
        return parseDeclarationEvents(node, context);
    }
    parseDeclarationSlots(node, context) {
        return parseDeclarationSlots(node, context);
    }
    parseDeclarationCSSProps(node, context) {
        return parseDeclarationCSSProps(node, context);
    }
}
